src/gui/qgsmaptoolcapture.h

```diff
@@ -255,10 +263,10 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
     bool mCapturing = false;
 
     //! Rubber band for polylines and polygons
-    QgsRubberBand *mRubberBand = nullptr;
+    std::unique_ptr<QgsRubberBand> mRubberBand;
```

Comment:
Aren't these owned by the canvas?

Reply:
You are right, the view (canvas) takes ownership of the item (rubberband).
But, the destructor of the maptool calls `stopCapturing()` which will delete the rubberbands already. Always.
This looks to me as if the map tool pretty much assumes ownership. And it's just safe because the canvas only gets deleted after the tools. I'm not sure how to best handle this situation.

I think leaving this as a unique_ptr is ok because it doesn't change the status quo but still makes memory management in here robust at no additional risks compared to the the current implementation.

But if you have a better idea, speak up!

Reply:
Well- maybe. 

What about a new unique_ptr type which uses a custom deleter and stores only a QPointer reference to an object? Then it'd be safe to use with parented QObjects...

Reply:
Sounds like a good plan. Do you have an approach implementation wise in mind?

1. Is basing it on unique_ptr possible? I think unique_ptr wants a pointer. Or can it be modified to manage a `QPointer<T>` object instead?
2. Alternatively a QPointer-based implementation could be done that deletes the managed object in the destructor. But I guess there all the move semantics etc. would need to be reimplemented and we wouldn't make use of lib-std implementations.

Number 2. would certainly be possible (I even have a draft ready). I don't see a way to get 1. even though I'd prefer it.

Reply:
Hmmmm. Maybe if 1 stores a pointer to a QPointer? It's a bit ugly though.

Reply:
Yes, Xzibit has put it this way:
![image](https://user-images.githubusercontent.com/588407/59195834-4a04f800-8b8d-11e9-8d8d-3b01f6f703f9.png)

Reply:
I would prefer not to use unique_ptr for any QObject-based pointers. I have already seen some crashes because of that (getting deleted by parent before getting deleted by unique_ptr)

Reply:
@wonder-sk do you have a smart idea on how to deal with scoped QObjects? It's a real world problem with no real solution so far.

Reply:
No smart ideas here... I would use QPointer and just keep a comment near declaration whether the object is parented (could, but does not need to be deleted) or not (must be deleted). Still better than odd crashes due to double delete (or early delete due to .reset() instead of deleteLater()) with unique_ptr.

Reply:
That's certainly safer w.r.t double delete. But riskier w.r.t leaks because of the lack of the "scoped" aspect.

I'll come up with a proposal for a scoped QPtr.